### DISCUSSION TITLE
Get BytesIO / file-like content from response

### DISCUSSION THREAD

**Discussion by davetapley** [encode/httpx#3028] [Ideas]
I'm using [`pandas.read_parquet`](https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html) and I need to wrap `content` in a `BytesIO` like this:

```py
response = await client.get(url)
df = read_parquet(BytesIO(response.content))
```

Is there a 'better' way to do this and / or could support for it be added?

I see [Streaming responses](https://www.python-httpx.org/async/#streaming-responses), but it makes no mention of `BytesIO` 🤔

**Comment by lovelydinosaur** [comment]
I think this is interesting, yep.

We have this API because we took a design call (somewhere?) that [the interface with `httpcore` should be an iterator](https://www.encode.io/httpcore/extensions/). I think it just seemed like the simplest of all possible options, so that's what we ended up with.

However...

* It's easy to take a `.read()` interface and and an iterator interface on top.
* It's not easy to take an iterator interface and add a `read()` interface on top.
* Having iterators through the stack potentially adds a bit of overhead, vs. the simpler `.read()` call chain.

So...

A sensible thing to do would be to investigate this further. An initial starting point would probably be implementing `.read()` for the interface in `httpcore`, and determining if there's any performance differences for large downloads vs. the existing `__iter__`.

I'd be happy to help you or anyone else work through how to do this.

I do think this probably makes sense, since we could add `.read()` to our stream interface without removing the existing `__iter__` support. But we would want to start by looking at this just at the `httpcore` level.

**Comment by graingert** [comment]
The UK dbt made a wrapper for iterabales with httpx in mind https://github.com/uktrade/to-file-like-obj
